### PR TITLE
Relax UpdatePropertiesFromSamples check constraints

### PR DESCRIPTION
**Pull request by JanuszL**
- UpdatePropertiesFromSamples checks if all tensors
  in a batch have equal properties. In some cases, when
  the tensor is empty it doesn't make much sense - like the device_id
  or pinnes level. This PR relaxes these constraints

## Category:
**Bug fix** (*non-breaking change which fixes an issue*)

## Description:
- UpdatePropertiesFromSamples checks if all tensors
  in a batch have equal properties. In some cases, when
  the tensor is empty it doesn't make much sense - like the device_id
  or pinnes level. This PR relaxes these constraints

### Affected modules and functionalities:
- tensor_list

### Key points relevant for the review:
- what should be checked for empty tensors

### Tests:

- [ ] Existing tests apply
- [x] New tests added
  - [ ] Python tests
  - [x] GTests
    - TensorListVariableBatchSizeTest_UpdatePropertiesFromSamples_Test
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [x] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

**Comment by JanuszL** [comment]
!build

File: dali/pipeline/data/tensor_list.cc [hunk 1]
```diff
@@ -1103,14 +1103,19 @@ void TensorList<Backend>::UpdatePropertiesFromSamples(bool contiguous) {
...
-  type_ = tensors_[0].type_info();
-  sample_dim_ = tensors_[0].shape().sample_dim();
-  shape_.resize(curr_num_tensors_, sample_dim_);
-  layout_ = tensors_[0].GetMeta().GetLayout();
-  pinned_ = tensors_[0].is_pinned();
-  order_ = tensors_[0].order();
-  device_ = tensors_[0].device_id();
-  contiguous_buffer_.set_order(order_);
+  for (int i = 0; i < curr_num_tensors_; i++) {
+    // if tensor is empty it can be uninitialized so we don't really care about its properties
```
Comment (klecki) [inline dali/pipeline/data/tensor_list.cc:1107]: What if we detected the first non-empty one, and propagated the setup to the empty ones?

File: dali/pipeline/data/tensor_list.cc [hunk 2]
```diff
@@ -1121,6 +1126,8 @@ void TensorList<Backend>::UpdatePropertiesFromSamples(bool contiguous) {
     DALI_ENFORCE(GetLayout() == tensors_[i].GetLayout(),
                  make_string("Samples must have the same layout, expected: ", GetLayout(),
                              " got: ", tensors_[i].GetLayout(), " at ", i, "."));
+    // if tensor is empty, its pinned status, order and device_id is rather irrelevant
+    if (tensors_[i].nbytes() == 0) continue;
```
Comment (klecki) [inline dali/pipeline/data/tensor_list.cc:1130]: And we just kept checking all the samples after such propagation.

Comment (JanuszL) [reply dali/pipeline/data/tensor_list.cc]: Done

**Comment by JanuszL** [comment]
!build

**Comment by JanuszL** [comment]
!build

**Review by klecki — APPROVED** [review]

**Review by jantonguirao — APPROVED** [review]

**Pull request merged**